enhancements/multi-arch/multiarch-manager-operator.md

```diff
@@ -0,0 +1,703 @@
+  network errors, ...), the scheduling gate is removed, and no changes are made
+  to the pod. In this case, the scheduling will only be architecture-aware if
+  the user submitting the pod defines the affinity predicates explicitly.
+
+### API Extensions
+
+#### Mutating Admission Webhook and Scheduling Gate 
+The multiarch-manager-operator reconciles not-yet-scheduled pods using the
+scheduling readiness[^1] and pod mutable scheduling directives[^2] features.
+Since Kubernetes 1.27[^3] promoted those features to beta, no additional changes
+to the Openshift Cluster Operators are required to leverage them in the context
+of automatically setting the architecture-related node affinity of pods. As
+described in the Workflow Description section, the
+`SchedulingGateMutatingAdmissionWebhook` changes a pod at admission time. It
+adds the `multi-arch.openshift.io/arch-aware-placement` scheduling gate to newly
+created pods. The scheduling gate can only be added to pods at creation time.
+The `PodPlacementController` is responsible for deleting the scheduling gate
+from the pod after inspecting the architecture of the pod containers' images and
+patching the node affinity set in the pod. The controller inspecting the image
+has access to any secrets in the cluster and builds the credentials to access
+the registries consuming the inspected pod's `imagePullSecret` field.  The
+controller will also read the image content source policy objects once the
+support for restricted network scenarios are implemented. 
+
+#### Required RBAC permissions
+
+The service account running the operator components will need the following RBAC
+permissions:
+
+- Deployment (namespace-scoped)
+  - Watch, List, Get, Create, Update, Delete
+- Pod (cluster-scoped)
+  - Watch, List, Get, Update
+- Secrets (cluster-scoped)
+  - Watch, List, Get
+- ConfigMaps (namespace-scoped)
+  - Watch, List, Get
+- PodPlacementConfig CRD
+  - Watch, List, Get, Update
+- Imagestreams/layers (required for internal registry access)
+  - Get
+
+#### PodPlacementConfig CR
+The configuration CR of the operator allows the admins to set the following:
+- a list of `insecureRegistries` to skip the TLS verification;
+- a list of `certificateAuthorityRefs` listing configMaps made of key-value
+  pairs containing the trusted certificates for the registries that use
+  self-signed certificates. For each CA file, the key in the ConfigMap is the
+  hostname of the registry in the `hostname[..port]` format, in compliance with
+  the `openshift-config/registry-cas` configMap leveraged by the Build APIs for
+  similar reasons. By default, the config CR will import the CAs in the above
+  mentioned configMap, if any.
+- a `skipTLSVerification` flag to completely disable the verification of the TLS
+  certificates while inspecting the pods. Although the reduced security, this
+  might guarantee higher performance when patching the nodeAffinity.
```

Comment:
Hi Ben, the priorities you described drive the phased approach we described.

We would also like this optional operator to be config-driven so that users can leverage it on other Kubernetes distributions. If we want to pre-render and build an automated OCP-specific configuration, we might target it phase by phase and add this implementation to the configuration controller. 

Do you have any suggestions for links (doc/code) to be sure we'll respect the alignment you are claiming wrt CRIO?

Reply:
i suggest you sync w/ @mrunalp or whoever he delegates to ensure alignment w/ kublet/CRIO image pull config+behavior.

Reply:
after some digging and research, a new section has been added around how we are going to deal with mirror registries and image verification(without requiring explicit configuration to be passed for each of them). I will also consult with Mrunal on this.

Reply:
Yeah, we need to ensure this. cc: @mtrmac and @saschagrunert

Reply:
> Although the reduced security, this might guarantee higher performance when patching the nodeAffinity.

I completely, and in the strongest terms, reject the premise that disabling TLS is a possible way to get acceptable performance.

TLS is what protects users’ pull secrets when contacting registries.

If this feature can not work well enough while enforcing TLS, then we simply don’t have a workable implementation.

(I would also note that I’m skeptical that this is really a major performance concern. To my knowledge every single connection to the API server, and between the API server / etcd, and between etcd nodes, uses TLS, and I’ve never heard disabling TSL would even be considered as an option to improve performance.

And especially with some persistent connections / connection pooling / … I would expect TLS session setup to be pretty negligible. Sure, that might force such features to be moved into phase 1, but again, if this were not performant enough while enforcing TLS, it would simply not be performant enough.

---

If you want to start with a phase 0 that only works with a specific list of public registries like `docker.io` and `quay.io`, and _explicitly refuses_ to use pull secrets (which would, BTW, exclude use of this feature for OCP release content), \*shrug\*, it would be a way to experiment.

But no version of this software that sends users’ pull secrets while disabling TLS can ever be released in the product.

Reply:
@mtrmac agreed and even we had reservations about doing this but we had thought that with a phased approach we could tackle it later - but makes sense. we will hash out the details and update the proposal shortly. I agree that performance is something we still have to quantify and evaluate before we make sure it is indeed a concern.

Reply:
Thanks. The recent TLS verification updates LGTM.

Reply:
@bparees could you PTAL the updates? all the concerns should be addressed.

Reply:
resolving this thread as the concerns have been addressed.